bpf/cachingmap/caching_map_test.go

```diff
@@ -0,0 +1,369 @@
+// Copyright (c) 2021 Tigera, Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cachingmap
+
+import (
+	"fmt"
+	"testing"
+
+	. "github.com/onsi/gomega"
+	"github.com/sirupsen/logrus"
+
+	"github.com/projectcalico/felix/bpf"
+	"github.com/projectcalico/felix/bpf/mock"
+	"github.com/projectcalico/felix/logutils"
+)
+
+func init() {
+	logutils.ConfigureEarlyLogging()
+	logrus.SetLevel(logrus.DebugLevel)
+}
+
+var cachingMapParams = bpf.MapParameters{
+	Name:      "mock-map",
+	KeySize:   2,
+	ValueSize: 4,
+}
+
+// TestCachingMap_Empty verifies loading of an empty map with no changes queued.
+func TestCachingMap_Empty(t *testing.T) {
+	mockMap, cm := setupCachingMapTest(t)
+	err := cm.ApplyAllChanges()
+	Expect(err).NotTo(HaveOccurred())
+	Expect(mockMap.Contents).To(BeEmpty())
+}
+
+
+var ErrFail = fmt.Errorf("fail")
+
+// TestCachingMap_Errors tests returning of errors from the underlying map.
+func TestCachingMap_Errors(t *testing.T) {
+	mockMap, cm := setupCachingMapTest(t)
+	mockMap.IterErr = ErrFail
+	err := cm.ApplyAllChanges()
+	Expect(err).To(HaveOccurred())
+
+	// Failure should have cleared the cache again so next Apply should see this new entry.
+	mockMap.Contents = map[string]string{
+		string([]byte{1, 1}): string([]byte{1, 2, 4, 3}),
+	}
+	mockMap.IterErr = nil
+	err = cm.ApplyAllChanges()
+	Expect(err).NotTo(HaveOccurred())
+	Expect(mockMap.Contents).To(BeEmpty())
+
+	// Now check errors on update
+	cm.SetDesiredState([]byte{1, 1}, []byte{1, 2, 4, 4})
+	mockMap.UpdateErr = ErrFail
+	err = cm.ApplyAllChanges()
+	Expect(err).To(HaveOccurred())
+
+	// And then success
+	mockMap.UpdateErr = nil
+	err = cm.ApplyAllChanges()
+	Expect(err).NotTo(HaveOccurred())
+	Expect(mockMap.Contents).To(Equal(map[string]string{
+		string([]byte{1, 1}): string([]byte{1, 2, 4, 4}),
+	}))
+
+	// And delete.
+	mockMap.DeleteErr = ErrFail
+	cm.DeleteAll()
+	err = cm.ApplyAllChanges()
+	Expect(err).To(HaveOccurred())
```

Comment:
deleting shouldn't happen here?

we can still have half updated map if error happens, right? But if we reload the cache, the desired state is still there and the operation that faild would be still pending. Do I get it right?

Reply:
Yeah, if the delete fails then we don't update pendingDeletions or cacheOfDataplane so we'll try again next time.

Reloading the cache should always bring cacheOfDataplane and the pendingXXX sets into sync.